### DISCUSSION TITLE
Change Request: Improve RuleTester error message for fixer output assertion

### DISCUSSION THREAD

**Discussion by adidahiya** [eslint/eslint#16247] [Q&A]
### ESLint version

8.21.0

### What problem do you want to solve?

I'm writing tests for my custom rules and I would like to see better output for the fix output assertions. Right now I just see "Output is incorrect." It seems like this problem was raised previously in #5917 and possibly fixed, but now it has regressed. Looking at the code, there is no diff printed here:

https://github.com/eslint/eslint/blob/9f5a75206675f2404ad6733640cf05ab44b02274/lib/rule-tester/rule-tester.js#L1005

### What do you think is the correct solution?

Print a pretty diff of the expected vs. actual output after lint fixes are applied.

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by bmish** [comment]
`assert.strictEqual()` should automatically show the diff between the actual/expected values so if it's not then something must be hiding that.

**Comment by adidahiya** [comment]
Here's how I'm running RuleTester:

https://github.com/palantir/blueprint/blob/3f904f0077ae055a845982c064e86775cb2c1abb/packages/eslint-plugin/test/no-deprecated-type-references.test.ts#L25

Note that it's running via typescript-eslint, but that implementation just passes straight through to ESLint's RuleTester:

https://github.com/typescript-eslint/typescript-eslint/blob/39829c01906f326fec94e9b3a5fdb1730eb02002/packages/utils/src/ts-eslint/RuleTester.ts#L189

**Comment by nzakas** [comment]
The assertion will print out a diff. This is likely caused by your testing setup swallowing some output.

**Comment by mdjermanovic** [comment]
@adidahiya how do you run the tests, directly or with a test framework? `assert.strictEqual` throws an error object with properties `expected` and `actual`, and then test frameworks use these values to display the diff. For example, this is how it looks like with `mocha` by default:

![image](https://user-images.githubusercontent.com/44349756/187045309-6e8ee78e-076a-42fa-b930-8e111a8710fe.png)

`mocha` with its `--inline-diffs` flag:

![image](https://user-images.githubusercontent.com/44349756/187045488-24447d18-976f-49d2-95cc-f9747352be25.png)


https://mochajs.org/#diffs

**Reply by adidahiya** [reply]
@mdjermanovic I'm running my tests through `ts-node`: https://github.com/palantir/blueprint/blob/3f904f0077ae055a845982c064e86775cb2c1abb/packages/eslint-plugin/package.json#L12

The test scripts run `{ RuleTester } from "@typescript-eslint/utils"`: https://github.com/palantir/blueprint/blob/bc7af39969cd2944beacfb75bca5606184ca2718/packages/eslint-plugin/test/no-deprecated-type-references.test.ts#L25

I'm going to try switching to Jest or Mocha and see if that fixes the console output...

**Reply by adidahiya** [reply]
I tried two things so far: upgrading `ts-node` and switching to just using `node` (with a pre-test-run `tsc` compilation step). Both produced the following diff output. It's not as pretty as Mocha's but it's something:

```
/Users/adahiya/dev/oss/blueprint/node_modules/eslint/lib/rule-tester/rule-tester.js:362
        throw err;
        ^

AssertionError [ERR_ASSERTION]: Output is incorrect.
    at testInvalidTemplate (/Users/adahiya/dev/oss/blueprint/node_modules/eslint/lib/rule-tester/rule-tester.js:1005:28)
    at Function.<anonymous> (/Users/adahiya/dev/oss/blueprint/node_modules/eslint/lib/rule-tester/rule-tester.js:1039:29)
    at Function.itDefaultHandler (/Users/adahiya/dev/oss/blueprint/node_modules/eslint/lib/rule-tester/rule-tester.js:357:23)
    at /Users/adahiya/dev/oss/blueprint/node_modules/eslint/lib/rule-tester/rule-tester.js:1036:69
    at Array.forEach (<anonymous>)
    at Function.<anonymous> (/Users/adahiya/dev/oss/blueprint/node_modules/eslint/lib/rule-tester/rule-tester.js:1035:30)
    at Function.describeDefaultHandler (/Users/adahiya/dev/oss/blueprint/node_modules/eslint/lib/rule-tester/rule-tester.js:374:19)
    at Function.<anonymous> (/Users/adahiya/dev/oss/blueprint/node_modules/eslint/lib/rule-tester/rule-tester.js:1034:30)
    at Function.describeDefaultHandler (/Users/adahiya/dev/oss/blueprint/node_modules/eslint/lib/rule-tester/rule-tester.js:374:19)
    at RuleTester.run (/Users/adahiya/dev/oss/blueprint/node_modules/eslint/lib/rule-tester/rule-tester.js:1022:26) {
  generatedMessage: false,
  code: 'ERR_ASSERTION',
  actual: 'import { SelectProps } from "@blueprintjs/select";\n' +
    'const mySelectProps: SelectProps = { items: [] };',
  expected: 'import { SelectProps } from "@blueprintjs/select";\n' +
    'const mySelectProps: ISelectProps = { items: [] };',
  operator: 'strictEqual'
}
```

I guess if I want to see a prettier diff I'll have to install `mocha`.

**Reply by adidahiya** [reply]
Adding `mocha` to my test setup fixed the issue, now I see a pretty diff:

<img width="433" alt="image" src="https://user-images.githubusercontent.com/723999/187978241-a8b681e1-9791-41cc-b0e8-398ff9252d37.png">

I didn't have to write any `describe()` or `test()` blocks, the rule tester seems to work with mocha automatically.
